x/incentive/keeper/adapters/hard_borrow/adapter.go

```diff
@@ -0,0 +1,48 @@
+package hard_borrow
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/kava-labs/kava/x/incentive/types"
+)
+
+var _ types.SourceAdapter = SourceAdapter{}
+
+type SourceAdapter struct {
+	keeper types.HardKeeper
+}
+
+func NewSourceAdapter(keeper types.HardKeeper) SourceAdapter {
+	return SourceAdapter{
+		keeper: keeper,
+	}
+}
+
+func (f SourceAdapter) TotalSharesBySource(ctx sdk.Context, sourceID string) sdk.Dec {
+	coins, found := f.keeper.GetBorrowedCoins(ctx)
+	if !found {
+		return sdk.ZeroDec()
+	}
+
+	return coins.AmountOf(sourceID).ToDec()
+}
```

Comment:
Hard has an additional complexity that the deposit amounts don't directly map to source shares.

A user's hard supplies and borrows technically increase every block due to interest, so they can't directly be used as source shares (otherwise we'd have to call sync every block). Similar to incentive the interest is not added to the store objects every block, it's calculated retroactively when a user deposits/withdraws/borrows/repays, so `Borrow.Amount` isn't the precise borrowed amount (like how the `Claim.Rewards` field isn't the actual claim's current reward balance).

The current code instead uses the 'normalized' borrow/deposit amounts as source shares:

https://github.com/Kava-Labs/kava/blob/aa5e9f63823575114ff7e3817aa9b299f3293cff/x/incentive/keeper/rewards_borrow.go#L39-L64
https://github.com/Kava-Labs/kava/blob/aa5e9f63823575114ff7e3817aa9b299f3293cff/x/incentive/keeper/rewards_borrow.go#L95-L96